source/develop/release-management/features/external-network-provider.html.md

```diff
@@ -0,0 +1,525 @@
+
+* provide the statistics for the VM NIC attached to the external network
+
+#### after_get_caps
+Called after generating the host capabilites (host configuration) which are to be send to the engine.
+Network provider reponsibilities:
+
+* provider information about the host network configuration related to the external network.
+Depending on the implementation, the provider might want report the external network being
+attached to any of the host NICS, report the external network being attached to a virtual host
+NIC, hide some host NICs used by the external network implementation, or report any other implemented configuration.
+
+#### before_network_setup/after_network_setup
+Called before and after the network setup operation on the host.
+Network provider reponsibilities:
+
+* if needed, the network provider might prevent network setup actions which would interfere with the external networks on the host.
+
+
+## External provider reference implementation
+
+A reference implementation of an external provider is prepared to show how.
+The implementation show in the reference implementation is very simple (naive),
+it's purpose is to show where rather than how different parts of the
+functionality should be implemented.
+
+The reference implementation consists of two parts:
+
+* the REST API - implementing the external provider REST API
+
+* VIF driver - implementing the driver which connects VM NICs on the hosts
+
+
+
+## Features under discussion
+
+### Provider custom properties
+
+Custom properties for an external network provider will be added. The user will be able to add the custom properties when adding the provider. The changed UI will should look like this (best guess - this might still change):
+
+![](external_network_provider_custom_properties.png "fig:external_network_provider_custom_properties.png")
+
+These properties will be passed to the VIF driver upon pluging of a nic, the same way the "vnic_id" property
+is passed.
+These properties might also be passed to the VIF driver when unplugging a nic (to be decided).
+
+
+### Automatic VIF driver deployment
+
+In the later stages of the external network provider feature, support for automated deployment
+of VIF drivers during host deploy will be added. This will be similar to the current OpenStack Neutron 
+provider deployment.
+It should be possible to add more than one VIF driver automatically.
+
+![](external_network_provider_automated_deploy.png "fig:external_network_provider_automated_deploy.png")
+
+The id of the provider would then be passed each time a nic is provisioned to let the appropriate VIF driver
+handle the connecting action.
+
+The binaries of the VIF driver could be provided on the host as an rpm (available in the host repos), and would
+include an installation script which would then hook up the VIF driver to the hooks.
+
+![](external_network_provider_provider_from_repo.png "fig:external_network_provider_provider_from_repo.png")
```

Comment:
But please make sure that there's a section about the need to integrate
with ovirt-host-deploy
On Mar 24, 2016 4:20 PM, "mmirecki" <notifications@github.com> wrote:

> In
> source/develop/release-management/features/external-network-provider.html.md
> <https://github.com/oVirt/ovirt-site/pull/103#discussion_r57322570>:
>
> > +### Automatic VIF driver deployment
> > +
> > +In the later stages of the external network provider feature, support for automated deployment
> > +of VIF drivers during host deploy will be added. This will be similar to the current OpenStack Neutron
> > +provider deployment.
> > +It should be possible to add more than one VIF driver automatically.
> > +
> > +![](external_network_provider_automated_deploy.png "fig:external_network_provider_automated_deploy.png")
> > +
> > +The id of the provider would then be passed each time a nic is provisioned to let the appropriate VIF driver
> > +handle the connecting action.
> > +
> > +The binaries of the VIF driver could be provided on the host as an rpm (available in the host repos), and would
> > +include an installation script which would then hook up the VIF driver to the hooks.
> > +
> > +![](external_network_provider_provider_from_repo.png "fig:external_network_provider_provider_from_repo.png")
>
> Removing this section and image
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/oVirt/ovirt-site/pull/103/files/99c4e7f6f4c05cec4d7d28fd476219c0a0d7a7bb#r57322570>
>